assignments/session03/calculator.py

```diff
@@ -0,0 +1,144 @@
+            return result_template.format("/", str(args[0] / args[1]), str(args[0]), str(args[1]))
+        else:
+            return error_template.format("/", "You can't divide by zero!")
+    else:
+        raise NameError
+
+
+def calculator():
+    return "calculator"
+
+
+def application(environ, start_response):
+    headers = [("Content-type", "text/html")]
+    try:
+        querystr = environ.get('QUERY_STRING', None)
+        if querystr is None:
+            raise NameError
+        # querystr is everything after the question mark, NOT including it:
+        # blabbityblah=6&test=1
+        if querystr is "":
+            body = homepage_template
+        else:
+            func, args = resolve_path(querystr)
+            body = func(*args)
+        status = "200 OK"
+    except NameError:
+        status = "404 Not Found"
+        body = "<h1>Not Found</h1>"
+    except Exception as e:
+        status = "500 Internal Server Error"
+        body = "<h1>Internal Server Error</h1>" + str(e)
+    finally:
+        headers.append(('Content-length', str(len(body))))
+        start_response(status, headers)
+        return [body.encode('utf8')]
+
+
+def resolve_path(querystr):
+    ''' query string should either be blank (if going to the homepage), or
+    come in with two operands and an operator, x1, x2, and op.
+    If it doesn't match either of these conditions, raise a NameError.
+    returns a 2-tuple including the function to call (multiply(), add(),
+    etc.) and the arguments as a list.
+    Should probably check the query string before creating the dictionary
+    to make sure there are no duplicate parameters. Someday...'''
+    if querystr is "" or querystr is None:  # main calculator page
+        func = calculator
+        args = []
+    else:
+        # parse query string:
+        # Cris suggested using the urllib parsing tools here. Haven't had a chance
+        # to look at this, but seems like using them would make more sense and
+        # avoid potential errors (blank query string values, etc.).
+        query_elements_list = [x for x in querystr.split("&")]
+        query_dict = dict(map(str, m.split("=")) for m in query_elements_list)
+        # set default operation to 'add', and x1 and x2 to 0 to avoid errors...
+        # should be handled more elegantly.
+        operation = query_dict.get("op", 'add')
+        args = [int(query_dict.get("x1", 0)), int(query_dict.get("x2", 0))]
+        # choose the appropriate func based on 'op' key value.
+        # should these be put into a global dictionary? Would this be better so
+        # the program doesn't run through all the 'elif' statements each time thru?
+        if operation == 'multiply':
+            func = multiplication
+        elif operation == 'divide':
+            func = division
+        elif operation == 'add':
+            func = addition
+        elif operation == 'subtract':
+            func = subtraction
+        else:  # Error condition
```

Comment:
Notice how similar the name of the string you are looking for is to the name of your func? Is there anything you could do with this that would eliminate the if statement?